docs: ADR-005 Liquidation (#239)
umee-network/umee: docs/architecture/ADR-005-liquidation.md

# ADR 005: Liquidation

## Changelog

- November 19, 2021: Initial Draft (@toteki)

## Status

Proposed

## Context

When borrowers on Umee exceed their borrow limit due to interest accrual or asset price fluctuations, their positions become eligible for liquidation by other users.

Such users (Liquidators) pay off part of all of the borrower's loan, in exchange for a value of collateral equal to the amount paid off plus an additional percentage (the liquidation incentive).

We must build the features necessary for liquidators to continuously look out for liquidation opportunities, then carry out chosen liquidations.

Additional parameters will be required which define the liquidation incentive and any other new variables.

## Decision

> Liquidation will require one message type (`MsgLiquidate`), one per-token parameter (`LiquidationIncentive`), and two global parameters (`MinimumCloseFactor` and `CompleteLiquidationThreshold`).

Reviewer: nit: I would just use `,` instead of `(...)`

There is no event type for when a borrower becomes a valid liquidation target, nor a list of valid targets stored in the module. Liquidators will have to use an off-chain tool to query their nodes periodically.

## Detailed Design

> A function `IsLiquidationEligible(borrowerAddr)` can be created to determine if a borrower is currently exceeding their borrow limit. Any liquidation attempt against a borrower not over their limit should fail.

Reviewer: ```suggestion
A function `IsLiquidationEligible(borrowerAddr)` can be created to determine if a borrower is currently exceeding their borrow limit. Any liquidation attempt against a borrower not over their limit will fail.
```

> A borrower's total borrower value (expressed in USD) can be computed from their total borrowed tokens and a price oracle.

Reviewer: ```suggestion
A borrower's total borrowed value (expressed in USD) can be computed from their total borrowed tokens and the `x/oracle` price oracle module.
```

> Their borrow limit is calculated similarly using borrower uToken balances, borrower collateral settings, current uToken exchange rates, and token collateral weights.

Reviewer: ```suggestion
Their borrow limit is calculated similarly using the borrower's uToken balance, their collateral settings, current uToken exchange rates, and token collateral weights.
```

> ### Basic Message Type

Reviewer: ```suggestion
### Message Types
```

To implement the liquidation functionality of the Asset Facility, one message type is required:

```go

> // MsgLiquidate - a liquidator targets a specific borrower, asset type, and collateral type for complete or partial liquidation

Reviewer: nit: let's wrap the godoc at 80 chars.

type MsgLiquidate struct {
  Liquidator    sdk.AccAddress
  Borrower      sdk.AccAddress
  Repayment     sdk.Coin // denom + amount
  RewardDenom   string
}
```

Repayment's denom is the borrowed asset denom to be repaid (because the borrower may have multiple open borrows). It is always a base asset type (not a uToken).
Its amount is the maximum amount of asset the liquidator is willing to repay. This field enables partial liquidation.

> RewardDenom is the collateral type which the liquidator will recieve in exchange for repaying the borrower's loan. It is always a uToken denomination.

Reviewer: The liquidator can pick any token used as collateral in the protocol, right? is it possible for one to run out? Will there be races or highly preferable tokens chosen? If so, would that be an issue?

Author: Any token that the borrower themselves has lent and marked as collateral.

So if I'm a borrower and I've lent `ETH`, `uAtom`, and `uOsmo`, and enabled only `uAtom` and `uOsmo` as collateral, then a liquidator targeting my address has the option of receiving `u/uAtom` and `u/uOsmo` as a reward. Those rewards come directly out of my balance, so there is no highly contested system pool.

Reviewer: Got it. Thanks!

It is necessary that messages be signed by the liquidator's account. Thus the method `GetSigners` should return the `Liquidator` address for the message type above.

### Partial Liquidation

It is possible for the amount repaid by a liquidator to be less than the borrower's total borrow position in the target denom. Such partial liquidations can succeed without issue, reducing the borrowed amount by `Repayment.Amount` and rewarding collateral proportional to the amount repaid.

Additionally, the `Repayment.Amount` found in `MsgLiquidate` functions as a maximum the liquidator is willing to repay. There are multiple factors that may reduce the actual repayment below this amount:

- Borrowed amount of `Repayment.Denom` is below `Repayment.Amount`
- Insufficient `RewardDenom` collateral to match repaid value plus liquidation incentive
- Parameter-based maximum, e.g. if a parameter `LiquidationCap = 0.50` caps liquidated value in a single transaction at 50% of the total borrowed value.

> In the above scenarios, the `MsgLiquidate` should succeed with the maximum amount allowed by the conditions, rather than fail outright.

Reviewer: We need to absolutely ensure that we emit useful events in the `MsgLiquidate` message handler. Specifically, in such a case where the full repayment is not met, the events should allude to what was actually repaid and how much. This will be useful for tracking purposes for liquidators ;-)

Author: Noted - when we get to events, all denominations and amounts involved in various transactions will be tracked.

### Token Parameters

In order to incentivize liquidators to target certain collateral types for liquidation first, the token parameter `LiquidationIncentive` is used.

When a `MsgLiquidate` causes liquidation to occur, the liquidator receives collateral equal to (100% + `RewardDenom.LiquidationIncentive`) of the repaid value worth of collateral.

For example, if the liquidation incentive for `uatom` is `0.15`, then the liquidator receives `u/uatom` collateral worth 115% of the borrowed base assets they repaid. The denom of the base assets does not affect this calculation.

### Calculating Liquidation Amounts

When a `MsgLiquidate` is received, the `x/leverage` module must determine if the targeted borrow address is eligible for liquidation.

```go
    // from MsgLiquidate (liquidatorAddr, borrowerAddr, repayDenom, repayAmount, rewardDenom)

>     borrowed := GetTotalBorrows(borrowerAddr)

Reviewer: We need to do this per-token right? Since `GetTotalBorrows` returns all borrowed tokens.

Author: GetTotalBorrows returns an `sdk.Coins` with all borrowed token types. We sum the value of all those coins when computing total borrowed value (so if we only build oracle with a single-denom `GetValue`, then we will have to call it once for each denom in the `sdk.Coins`).

Reviewer: Ok, that's not super clear from this pseudo-code but I don't think it's a big deal. Feel free to modify it and make it clearer at your own discretion 👍

    liquidationIncentive, closeFactor := GetLiquidationParameters(rewardDenom, borrowValue, collateralValue)
```

The liquidation incentive is the bonus collateral received when a liquidator repays a borrow position
(e.g. incentive=`0.2` means liquidator receives 120% the value of their repayment back in collateral).

The close factor is the portion of a borrow position eligible for liquidation in this single liquidation event.

See _Dynamic Liquidation Parameters_ section at the bottom of this document.

Once parameters are fetched, the final liquidation amounts (repayment and reward) must be calculated.

```go
   // Repayment cannot exceed liquidator's available balance
   liquidatorBalance := GetBalance(liquidatorAddr).Amount(repayDenom)
   repayAmount = Min(repayAmount, liquidatorBalance)

   // Repayment cannot exceed borrowed value * close factor
   repayValue := oracle.Value(sdk.NewCoin(repayDenom, repayAmount)) // price oracle
   if repayValue > borrowValue * closeFactor {
     partial := (borrowValue * closeFactor) / repayValue
     repayAmount = repayAmount * partial
   }

   // Collateral reward assuming current repayment amount

>    rewardAmount := oracle.EquivalentValue(repayDenom, repayAmount, rewardDenom.ToBaseAssetDenom) // price oracle

Reviewer: what does this mean?

Author: Another helper function that either the oracle will have, or we should build into our keeper (probably the latter, now that I think of it).

But the function should do "for a given denom and amount, find the amount in a new denom which would have equivalent value".

I'll clarify that in the doc since I didn't do so originally.

Reviewer: Got it. I don't think think that needs to be in the oracle keeper though, right? What it sounds like we need to do is get the price P1 of the denom D1, get the price P2 of denom D2 and determine how much of D2 we need. I.e. this sounds like just a private helper function we can define right in the leverage module, meaning no new APIs to x/oracle.

>    rewardAmount := oracle.EquivalentValue(repayDenom, repayAmount, rewardDenom.ToBaseAssetDenom) // price oracle

Author: ```suggestion
   // given repay denom and amount, calculate the amount of rewardDenom that would have equal value
   rewardAmount := oracle.EquivalentValue(repayDenom, repayAmount, rewardDenom.ToBaseAssetDenom) // price oracle
```

# ADR 005: Liquidation

## Changelog

- November 19, 2021: Initial Draft (@toteki)

## Status

Proposed

## Context

When borrowers on Umee exceed their borrow limit due to interest accrual or asset price fluctuations, their positions become eligible for liquidation.

Third party liquidators pay off part of all of the borrower's loan, in exchange for a value of collateral equal to the amount paid off plus an additional percentage (the liquidation incentive).

We must build the features necessary for liquidators to continuously look out for liquidation opportunities, then carry out chosen liquidations.

Additional parameters will be required which define the liquidation incentive and any other new variables.

## Decision

> Liquidation will require one message type (`MsgLiquidate`), one per-token parameter (`LiquidationIncentive`), and two global parameters (`MinimumCloseFactor` and `CompleteLiquidationThreshold`).

Author: ```suggestion
Liquidation will require one message type `MsgLiquidate`, one per-token parameter `LiquidationIncentive`, and two global parameters `MinimumCloseFactor` and `CompleteLiquidationThreshold`.
```

> Dynamic close factor
>
> Close factor ranges from `0.0 = MinimumCloseFactor` to 1.0 when the borrower is between 0% and `20% = CompleteLiquidationThreshold` over borrow limit, then stays at 1.0
>
> | Borrow Limit (BL) |Borrowed Value (BV) | BV / BL | Close Factor |
> | - | - | - | - |
> | 100 | 100.1 | 1.001 | 0.005 |
> | 100 | 102 | 1.02 | 0.1 |
> | 100 | 110 | 1.1 | 0.5 |
> | 100 | 130 | 1.2 | 1.0 |
> | 100 | 140 | 1.4 | 1.0 |

The Dynamic Close Factor takes advantage of market forces to reduce excessive collateral selloffs, by reducing the portion of collateral initially eligible for liquidation.
Liquidators would have the change to liquidate smaller portions of the borrow if profitable and bring the position back into health.
Otherwise, close factor would continue to increase as the borrow accrues interest.

This also allows borrows to be liquidated completely in one transaction, once they are serverely over their borrow limit.

The `LiquidationIncentive` parameter can be any value, varying from token to token, without affecting the close factor.

> 

Author: ```suggestion

## Querying for Eligible Liquidation Targets

The offchain liquidation tool employed by liquidators will need several queries to be supported in order to periodically scan for and act on liquidation opportunities:

- GetLiquidationTargets: Return a list of all borrower addresses that are over their borrow limits
- GetTotalBorrows(borrower): Returns an `sdk.Coins` containing all assets borrowed by a single borrower
- GetTotalCollateral(borrower): Returns an `sdk.Coins` containing all uTokens in borrower's balance that are enabled as collateral

In addition to the above, the liquidation tool should be able to read any global or per-token parameters in order to finish calculating borrow limits and potential liquidation rewards.

```

- Parameter-based maximum, e.g. if a parameter `LiquidationCap = 0.50` caps liquidated value in a single transaction at 50% of the total borrowed value.

In the above scenarios, the `MsgLiquidate` should succeed with the maximum amount allowed by the conditions, rather than fail outright.

### Token Parameters

In order to incentivize liquidators to target certain collateral types for liquidation first, the token parameter `LiquidationIncentive` is used.

When a `MsgLiquidate` causes liquidation to occur, the liquidator receives collateral equal to (100% + `RewardDenom.LiquidationIncentive`) of the repaid value worth of collateral.

For example, if the liquidation incentive for `uatom` is `0.15`, then the liquidator receives `u/uatom` collateral worth 115% of the borrowed base assets they repaid. The denom of the base assets does not affect this calculation.

### Calculating Liquidation Amounts

When a `MsgLiquidate` is received, the `x/leverage` module must determine if the targeted borrow address is eligible for liquidation.

```go
    // from MsgLiquidate (liquidatorAddr, borrowerAddr, repayDenom, repayAmount, rewardDenom)

    borrowed := GetTotalBorrows(borrowerAddr)

>     borrowValue := oracle.TotalValue(borrowed) // price oracle

Author: ```suggestion
    // sum over the value of every denom in the total borrowed sdk.Coins
    borrowValue := oracle.TotalValue(borrowed) // price oracle
```

    collateral := GetCollateralBalance(borrowerAddr)
    collateral = MultiplyByCollateralWeight(collateral)

>     collateralValue := oracle.TotalValue(collateral) // price oracle

Author: ```suggestion
    // sum over the value of every sdk.Coin in the total collateral sdk.Coins
    collateralValue := TotalValue(collateral) // price oracle
```

    // collateral denomination, borrowed value, and collateral value
    liquidationIncentive, closeFactor := GetLiquidationParameters(rewardDenom, borrowValue, collateralValue)
```

The liquidation incentive is the bonus collateral received when a liquidator repays a borrow position
(e.g. incentive=`0.2` means liquidator receives 120% the value of their repayment back in collateral).

The close factor is the portion of a borrow position eligible for liquidation in this single liquidation event.

See _Dynamic Liquidation Parameters_ section at the bottom of this document.

Once parameters are fetched, the final liquidation amounts (repayment and reward) must be calculated.

```go
   // Repayment cannot exceed liquidator's available balance
   liquidatorBalance := GetBalance(liquidatorAddr).Amount(repayDenom)
   repayAmount = Min(repayAmount, liquidatorBalance)

   // Repayment cannot exceed borrowed value * close factor
   repayValue := oracle.Value(sdk.NewCoin(repayDenom, repayAmount)) // price oracle
   if repayValue > borrowValue * closeFactor {
     partial := (borrowValue * closeFactor) / repayValue
     repayAmount = repayAmount * partial
   }

>    // Given repay denom and amount, calculate the amount of rewardDenom
>    // that would have equal value
>    rewardAmount := oracle.EquivalentValue(repayDenom, repayAmount, rewardDenom.ToBaseAssetDenom) // price oracle

Author: ```suggestion
   // Given repay denom and amount, calculate the amount of rewardDenom
   // that would have equal value
   rewardAmount := EquivalentValue(repayDenom, repayAmount, rewardDenom.ToBaseAssetDenom) // price oracle
```

Liquidators would have the change to liquidate smaller portions of the borrow if profitable and bring the position back into health.
Otherwise, close factor would continue to increase as the borrow accrues interest.

This also allows borrows to be liquidated completely in one transaction, once they arr serverely over their borrow limit.

The `LiquidationIncentive` parameter can be any value, including varying from token to token, without affecting this method.

> **++++++++++++++++++**
> **DECISIONS REQUIRED**
> **++++++++++++++++++**
> 
> Which strategy should we employ for dynamic liquidation parameters
> 
> a) No dynamic liquidation. Has parameters:
>    - Per-token `LiquidationIncentive`
>    - Global parameter `CloseFactor`
>
> b) Dynamic incentive (option 1), with parameters:
>    - Per-token `MinLiquidationIncentive`
>    - Per-token `MaxLiquidationIncentive`
>    - Global parameter `FullLiquidationIncentiveThreshold`
>    - Global parameter `CloseFactor`
>
> c) Dynamic close factor (option 2), with parameters:
>    - Per-token `LiquidationIncentive`
>    - Global parameter `MinimumCloseFactor`
>    - Global parameter `CompleteLiquidationThreshold`
>

> > Once this has a final decision, I can update the document appropriately.

Author: Decision required here. It discusses the dynamic liquidation strategies proposed in the alternatives section.

Author: Going with option 2 (meeting)
